src/nftfeed.sol

```diff
@@ -0,0 +1,160 @@
+    mapping (bytes32 => uint) public nftValues;
+    // nftID => risk
+    mapping (bytes32 => uint) public risk;
+
+    // risk => thresholdRatio
+    mapping (uint => uint) public thresholdRatio;
+    // risk => ceilingRatio
+    mapping (uint => uint) public ceilingRatio;
+
+
+    PileLike pile;
+    ShelfLike shelf;
+
+    constructor () public {
+        wards[msg.sender] = 1;
+    }
+
+    function init() public {
+        require(thresholdRatio[0] == 0);
+        // risk groups are pre-defined and should not change
+        // gas optimized initialization of risk groups
+
+        // default risk => 0
+        // thresholdRatio => 80%
+        // ceilingRatio => 60%
+        // interestRatePerSecond results  =>  5 % day
+        setRiskGroup(0, 8*10**26, 6*10**26, uint(1000000564701133626865910626));
+
+        // risk group  => 1
+        // thresholdRatio => 70%
+        // ceilingRatio => 50%
+        // interestRatePerSecond results  =>  12 % day
+        setRiskGroup(1, 7*10**26, 5*10**26, uint(1000001311675458706187136988));
+    }
+
+    /// sets the dependency to another contract
+    function depend(bytes32 contractName, address addr) external auth {
+        if (contractName == "pile") {pile = PileLike(addr);}
+        else if (contractName == "shelf") { shelf = ShelfLike(addr); }
+        else revert();
+    }
+
+    /// returns a unique id based on registry and tokenId
+    /// the nftID allows to define a risk group and an nft value
+    /// before a loan is issued
+    function nftID(address registry, uint tokenId) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(registry, tokenId));
+    }
+
+    function nftID(uint loan) public view returns (bytes32) {
+        (address registry, uint tokenId) = shelf.shelf(loan);
+        return nftID(registry, tokenId);
+    }
+
+    /// Admin -- Updates
+    function setRiskGroup(uint risk_, uint thresholdRatio_, uint ceilingRatio_, uint rate_) internal {
+        thresholdRatio[risk_] = thresholdRatio_;
+        ceilingRatio[risk_] = ceilingRatio_;
+        // the risk group is used as a rate id in the pile
+        pile.file("rate", risk_, rate_);
+    }
+
+    ///  -- Oracle Updates --
+
+    /// update the nft value
+    function update(bytes32 nftID_,  uint value) public auth {
+        nftValues[nftID_] = value;
+    }
+
+    /// update the nft value and change the risk group
+    function update(bytes32 nftID_, uint value, uint risk_) public auth {
+        require(thresholdRatio[risk_] != 0, "threshold for risk group not defined");
+
+        // change to new rate immediately in pile if a loan debt exists
+        uint loan = shelf.nftlookup(nftID_);
+        if (pile.pie(loan) != 0) {
+            pile.changeRate(loan, risk_);
+        }
+
+        risk[nftID_] = risk_;
+        nftValues[nftID_] = value;
+    }
+
+    // method is called by the pile to check the ceiling
+    function borrow(uint loan, uint amount) external auth {
```

Comment:
don't we also need a principal implementation that we are currently using per default?

Reply:
I added a principal implementation in my latest commit